### ISSUE TITLE
Action timing out when new node versions are released

### ISSUE THREAD

**Issue by ebwinters** [actions/setup-node#132]
I am using this in a job:
```yaml
name: Use Node.js ${{ matrix.node-version }}
uses: actions/setup-node@v1
  with:
    node-version: ${{ matrix.node-version }}
```
and `node-version` `13.x`, but my workflow takes 10+minutes just to complete use node. Any idea why or anyone else having this problem?

**Comment by Xetera** [comment]
Same here, It looks like I'm sometimes timing out and getting
> Premature close
> Waiting 16 seconds before trying again
> Premature close
> Waiting 18 seconds before trying again
> ##[error]Premature close

and sometimes it's working fine but still taking 7-11 minutes per install for some reason (12.x)

**Comment by ebwinters** [comment]
good to know other people are having issues - seriously stalling my pipeline 😞

**Comment by camelmasa** [comment]
Same here, We are using node 12.x.

**Comment by fishcharlie** [comment]
Same. You can see the workflow on an open source project experiencing this [here](https://github.com/dynamoosejs/dynamoose/pull/779/checks?check_run_id=563127394). From my experience this just started happening recently.

As you can see the setup node step took 7 minutes and 34 seconds 😱 

<img width="1106" alt="image" src="https://user-images.githubusercontent.com/860375/78520227-0fef8d00-7783-11ea-9db6-72201b933f7d.png">

**Comment by fishcharlie** [comment]
One of the **major** issues of this action is the fact that there is next to no logs that get printed. I brought this up in #90 for the version it's using. But this is another situation where **we need logs**!!

**Comment by jiyuhan** [comment]
This is a serious issue if one's pipeline doesn't have default timeout setup. This will cause a lot of people's CI to hang for up to 6 hours. This is not wasting MS' resource, but is wasting resource for other paying customers. Please fix this!

**Comment by boredland** [comment]
this did cost us a lot of money already. please have a look!

**Comment by matiaschileno** [comment]
We have the same problem. Today every pipeline is failing because of this. We are using Node.js 12.x.

**Comment by ssannidhi-palo** [comment]
Experiencing similar issues with Node 12.x version. The step is either timing out or takes up too much time. (~25 mins)

![image](https://user-images.githubusercontent.com/54425886/78541442-374c6580-7828-11ea-81e8-0ff0734bd91e.png)

**Comment by robbruce** [comment]
Cloned the repo and ran the unit tests, it would appear that the issue is with nodejs.org

```
::debug::Failed to download from "https://nodejs.org/dist/v10.16.0/node-v10.16.0-darwin-x64.tar.gz". Code(500) Message(Internal Server Error)
```

It's not consistent though, a re-run of tests can get past this issue.

**Comment by emalihin** [comment]
Tried this action for the first time last night and thought 5-10+ minutes setup is natural until i saw this issue. +1

**Comment by WesleyKlop** [comment]
Just browsing random pages on https://nodejs.org sometimes gives a 500 error as well, so it looks like this issue is indeed on their end

**Comment by defionscode** [comment]
It's being tracked in the node upstream repo https://github.com/nodejs/node/issues/32683

**Comment by iliyanyotov** [comment]
@defionscode Thanks for linking to the node repo issue.

![image](https://user-images.githubusercontent.com/24395456/78545907-3bb35700-7805-11ea-9ec1-96d60b0ccbd0.png) 😢

**Comment by soanaphikeste** [comment]
Depending on what you need, you might skip this action completely:

All runners have node 12.16.1 preinstalled (e.g. https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md, search for "Node.js")

**Comment by Raul6469** [comment]
Another alternative is to run your job in a container :
```yml
your_job:
  runs-on: ubuntu-latest
  container: node:10.9
```

**Comment by Benwick91** [comment]
I have to set the registry-url and the scope in my workflow. And I have the same issues since today. Last week it runs in 1-2 min, but today the action fails or it durates up to 20 min. :(

**Comment by 19h** [comment]
@Raul6469 won't work if you depend on the default dependencies in the Github containers like the AWS CLI. Although I think that's something one should plan half a man-day for..

**Comment by culshaw** [comment]
This definitely warrants a feature request to support mirrors

**Comment by konradpabjan** [comment]
cc @damccorm @ericsciple @bryanmacfarlane 

I'm seeing this issue as well (on Mac and Linux it's taking me 5+ minutes for this action to run, Windows for some reason seems fine)

**Comment by damccorm** [comment]
@konradpabjan Node itself is having an outage right now that is causing this - https://github.com/nodejs/node/issues/32683

**Comment by adriaanvanrossum** [comment]
Omitting this whole package made it super fast again:

```yml
name: Use Node.js ${{ matrix.node-version }}
uses: actions/setup-node@v1
  with:
    node-version: ${{ matrix.node-version }}
```

It will have node 12.x by default, so no need to use this package when you don't care about the Node.js version.

**Comment by mieszko4** [comment]
I had an issue with time out set up to 6h over month ago because my script hanged.
Yes, it ate up all my minutes for the month. **Who sets a default timeout to 6h?!**
For me it is a trap. I wrote email to github support hoping they could compensate me some minutes because 6h timeout default is not clearly written in the docs but they just redirected my to "Product Team" which did not reply to me since.
**Since then I set up all my jobs with a reasonable timeout.**

And now/today I am wasting my time and github minutes for trying to run builds multiple times hoping it will pass without `Unexpected HTTP response: 500` or `##[error]Premature close`. I am really mad at this point.

**Comment by swanson** [comment]
If you're finding this thread and blocked by timeout issues, two things you may want to try:

- Removing this step: base images now have *a version* of Node installed -- so if you don't specifically need a certain version (for doing asset-type stuff not running your whole app), try removing the step
- Setting a timeout: your builds will fail faster, not run up your minutes, etc -- add `timeout-minutes` to your jobs (try something like `1.5x` your normal runtime) https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes

**Comment by mieszko4** [comment]
For refereenc software installed on default on `runs-on`: https://help.github.com/en/actions/reference/software-installed-on-github-hosted-runners

**Comment by bryanmacfarlane** [comment]
@mieszko4 - yes, if you don't need or want to validate on a specific version or version spec (range) of node, then you can rely on the one that's globally installed on the machine image.

Note that the global version on the image will slide (13, then 14, etc.).   

The core value of setup-node is using a specific version I want to test my code against lts 8, 10, 12, and latest in a matrix in parallel.  

That said, we are fixing [this issue now](https://github.com/actions/virtual-environments/issues/42) so that you can have the best of both worlds.  You can lock to an LTS (12.x) and do matrix testing and never do a download.

If it's not matching a major LTS, then it will download.

So part 2 of the work (also starting now) is the build a cache (backed by CDN) of versions so when we do read through, it hits our cache.

I might add a flag in the action to force reading through and getting absolute latest if you need that since there will be some latency on both the images and the cache.

**Comment by dentuzhik** [comment]
> Note that the global version on the image will slide (13, then 14, etc.).

A bit off-topic, but wondering what's the motivation of sliding the node in non-LTS increments in the global images. Since node 8, I have had very few cases in production, when latest-greatest node was desirable (especially without significant amount of testing first). 

> The core value of setup-node is using a specific version I want to test my code against lts 8, 10, 12, and latest in a matrix in parallel.

Very nice clarification, thank you. I believe it could have some better emphasis in readme (there's a good usage example, but no why motivation).

> Thanks for the note about setting timeout. 1.5x your typical time is a good suggestion.

We also have been hit quite hard by this issue, and rigorous timeouts on every job indeed saved our a$$es in this case, otherwise we would certainly hit the limit of minutes quite fast.

\cc @Kerizer

**Comment by bryanmacfarlane** [comment]
@dentuzhik - I should have clarified that hosted images should be installing LTS globally so the real scenario would be sliding from 10 to 12 and then 12 to 14.

I'll update the readme with all of this but also note that one more variable is self-hosted.

**Comment by dentuzhik** [comment]
If there will be a PR, I would happily have a look at it 🙇

**Comment by amio** [comment]
Seems we already using `@actions/tool-cache` in master branch 🤔

https://github.com/actions/setup-node/blob/83c9f7a7df54d6b57455f7c57ac414f2ae5fb8de/src/installer.ts#L149

**Comment by bryanmacfarlane** [comment]
@amio - yes, this action downloads from the node distribution to the tool-cache folder.  It does if it can't match the semver version spec to what's already in the tool-cache folder.  In self-hosted runners, that cache fills up on use and gets faster over time.  In hosted runners, it depends on whether that cache is prepped with likely hits (e.g. latest of every LTS major version line).  Remember that hosted VMs are a fresh machine every time.  Thus the mention of [this issue above](https://github.com/actions/virtual-environments/issues/42) to prep that cache as part of our image gen process.  Doing that will mean virtual 0 second resolution of 8.x, 10.x, 12.x specs and no reliability issues.

Part 2 of the work refers to a different cache.  An off VM (CDN) where we cache all the node distributions so everyones workflows aren't hitting the node distribution directly (which is not a CDN).  We can also get better locality and more reliability from a CDN.  This is essentially what was [suggested in the node issue](https://github.com/nodejs/node/issues/32683#issuecomment-609747749).  This is not a trivial task but we're working on it now.  You can see the [toolkit start here](https://github.com/actions/toolkit/pull/382).  The vm gen to populate that cache isn't OSS yet.  It's early and we're sketching it out.

So essentially, it will be (1) check local VM cache (great for self hosted, prepped on hosted) then (2) if no match locally, download from CDN then (3) if still no match, fall back to node dist.   

Hope that helps clarify.

**Comment by 19h** [comment]
I think it would be beneficial if Github was opting to mirror and effectively the Node binaries here if it wants Github Actions to be taken even remotely serious for production use.

**Comment by 19h** [comment]
I think it would be beneficial if Github was opting to mirror and effectively host the Node binaries if it wants Github Actions to be taken even remotely serious for production use.

**Comment by mataide** [comment]
its possible to set `registry-url: https://registry.npmjs.org/` without `uses: actions/setup-node@v1` ?

**Comment by bryanmacfarlane** [comment]
Update: v2-beta is available which downloads from a releases cache for additional reliability.  It falls back to node dist on miss or failure.  We also added latest of each LTS directly on the image so if you reference a major version binding ('10', '12', etc.) it should be virtually instant with no reliability issues from downloading.

More here: https://github.com/actions/setup-node#v2-beta

Note that this is the planned solution to this problem.

**Comment by fishcharlie** [comment]
@bryanmacfarlane It looks like v2 has been released (not in beta). What is the status on this issue? Are there still more planned changes to improve the reliability here or does v2 resolve everything you have planned?

**Comment by dmitry-shibanov** [comment]
Hello @patrickk. Thank you for your report. Firstly the action will try use nodejs (LTS) from toolcache preinstalled on the hosted images. If the action do not match version it will try to download nodejs (LTS) from [node-versions](https://github.com/actions/node-versions). If the action does not find a version in the node-versions, it'll try to download it from https://nodejs.org/dist 

Could you please create an issue with public repository or repro steps ?

**Comment by patrickk** [comment]
> Hello @patrickk. Thank you for your report. Firstly the action will try use nodejs (LTS) from toolcache preinstalled on the hosted images. If the action do not match version it will try to download nodejs (LTS) from [node-versions](https://github.com/actions/node-versions). If the action does not find a version in the node-versions, it'll try to download it from https://nodejs.org/dist
> 
> Could you please create an issue with public repository or repro steps ?

sure @dmitry-shibanov! Thanks for your reply. 

Unfortunately I can't create a public repo since a lot of it is private but I can detail out the steps:

1. Workflow checks out code
2. `setup-node@v1` tries downloading `16.11.11` 
3. 500s


<img width="800" alt="Screen Shot 2022-03-01 at 2 49 42 PM" src="https://user-images.githubusercontent.com/17170535/156238369-c0f29dd7-2323-4b40-ac9f-af0b884d24c9.png">

<img width="1212" alt="Screen Shot 2022-02-28 at 5 36 14 PM" src="https://user-images.githubusercontent.com/17170535/156238520-4f4532d3-4dfb-408d-be0d-3c5c771e7b0c.png">

(there is a version discrepancy but I'd intermittently get 500s on nodejs.org's dist for various versions). 

Read somewhere that this usually happens with new releases for node, but saw that the latest release in dist url was for Feb 22, while the error above occurred Feb 28.


We were able to get past this by letting the action fail a bunch/letting it run for 5+ minutes, and one of them finally succeeded. 

Haven't been able to reproduce (also I updated our workflow files to use v2 as of this morning).
